nixos/modules/system/boot/stage-1-init.sh

```diff
@@ -82,22 +82,28 @@ waitDevice() {
     # great if we had a way to synchronously wait for them, but
     # alas...  So just wait for a few seconds for the device to
     # appear.
-    if test ! -e $device; then
-        echo -n "waiting for device $device to appear..."
-        try=20
-        while [ $try -gt 0 ]; do
-            sleep 1
-            # also re-try lvm activation now that new block devices might have appeared
-            lvm vgchange -ay
-            # and tell udev to create nodes for the new LVs
-            udevadm trigger --action=add
-            if test -e $device; then break; fi
-            echo -n "."
-            try=$((try - 1))
-        done
-        echo
-        [ $try -ne 0 ]
-    fi
+
+    local IFS=':' # separate multiple devices using field splitting
+
+    for device in $device
```

Comment:
if device is changing from a single item to multiple, it should probably be renamed to devices

Reply:
Why make this change anyway?

Reply:
Not sure about this `:` stuff. I frequently pass disks in by path, and those paths often/always have colons: `/dev/disk/by-path/pci-0000:04:00.0-nvme-1-part2` I'm afraid biting the bullet and passing it in explicitly as a list would be better.

Reply:
I was split about that, will change.
It's necessary because bcachefs filesystems can consist of multiple devices and are then only mountable like `/dev/sdx1:/dev/sdy1`. That then ends up in the device attribute of the filesystem. As far as I can tell the string is only used for information to the user and for mounting - and here, for checking for the device presence. Of course that's not a valid device, so we separate it into its constituents beforehand.

Reply:
Good catch. I'll test out how bcachefs behaves when trying to mount these paths before I try and determine a path forward.

Reply:
bcachefs seems to handle this case correctly.
Unfortunately, I'm having trouble finding / understanding the code that ensures that: https://evilpiepirate.org/git/bcachefs.git/tree/fs/bcachefs/fs.c?id=c0f1098d4d6e0003e91da47695bcd8f06bb60797#n1573
Ideally I would like to emulate the responsible logic instead of rolling my (probably flawed) own.

It seems to me we have to do the same parsing either way, be it in Nix (and passing the device list) or in the shell script.

Reply:
I tested bcachefs some more and it turns out you can't mount multidevs with colons in the device name. This means there is probably going to be some change from upstream to how multidevs are specified.
Considering that, I suggest only separating when `:` is followed by `/` for now and accepting the long wait times for the rare case that someone actually wants to mount a device using a path with a directory name ending in `:`.

Reply:
Good news, I talked to Kent (the bcachefs author) and apparently he is going to add a mount helper to bcachefs-tools, which means filesystems will be mountable by their UUID. This eliminates the need of any changes to ´stage-1-init.sh´ at all. When that lands, I'll try and add a conversion from the colon-separated device string to the UUID to ´nixos-generate-config´ (if multidevices can still be mounted like that then). We'll also have to wait for the timeouts until bcachefs is updated to that point, but that's a fair price to pay for leaving everything as is, I think.
Only if someone feels it would be advisable to put support for eliminating the timeout for multi-device bcachefs filesystems on USB storage devices would it be advisable to implement the discussed separate device list.